### DISCUSSION TITLE
Error building app when build variants are added - Android Studio

### DISCUSSION THREAD

**Discussion by murasu** [axmolengine/axmol#1508] [Q&A] [answered]
I have successfully 'ported' my app from Cocos2d-x 3.x to Axmol. It was truly a delight to work with a refreshed framework :)

I am facing a problem in Android Studio. After compiling and running my app, I added build variants to get different versions of the app (targeting different people in different regions).

I am able to compile the apk. However, the axslc assets are not getting copied into the package. In other words, the first part of the Gradle script `android.applicationVariants.all { variant ->  .. }` works. The files in `Content` are copied over. The second part that copies files from `${projectDir}/../build/runtime/axslc` is not getting executed at all. 

Before adding the variants, both parts run without any issues.

I'm not a Gradle expert. Appreciate any pointers on where I should look for solutions. Thanks.

**Comment by rh101** [comment]
How did you set up the variants? Are you able to provide a sample project with a similar variants/flavors set-up that reproduces the issue?

**Reply by murasu** [reply]
I will try and setup a sample project. That may be easier :)

**Reply by rh101** [reply]
Never mind! I found the issue. Give me a few minutes to figure out what you can do about it.

**Reply by murasu** [reply]
Awesome! Will look forward. Thank you!

**Reply by rh101** [reply, marked as answer]
At the end of the `app/build.gradle` file, look for this section:

```
project.afterEvaluate {
    if (tasks.findByName("externalNativeBuildDebug")) {
        copyDebugAxslcToAssets.dependsOn externalNativeBuildDebug
        compileDebugJavaWithJavac.dependsOn copyDebugAxslcToAssets
    }
    if (tasks.findByName("externalNativeBuildRelease")) {
        copyReleaseAxslcToAssets.dependsOn externalNativeBuildRelease
        compileReleaseJavaWithJavac.dependsOn copyReleaseAxslcToAssets
    }
}
```

You may need to update that section with your required variants. For example, if your variant is named "custom":
```
project.afterEvaluate {
    if (tasks.findByName("externalNativeBuildDebug")) {
        copyDebugAxslcToAssets.dependsOn externalNativeBuildDebug
        compileDebugJavaWithJavac.dependsOn copyDebugAxslcToAssets
    }
    if (tasks.findByName("externalNativeBuildRelease")) {
        copyReleaseAxslcToAssets.dependsOn externalNativeBuildRelease
        compileReleaseJavaWithJavac.dependsOn copyReleaseAxslcToAssets
    }

    // New section for the "custom" variant (note the naming)
    if (tasks.findByName("externalNativeBuildCustom")) {
        copyCustomAxslcToAssets.dependsOn externalNativeBuildCustom
        compileCustomJavaWithJavac.dependsOn copyCustomAxslcToAssets
    }
}
```

I'm not sure if this can somehow be automated based on the variants, but for now, see if that works for you.

**Reply by murasu** [reply]
This works! I had to append 'Debug' and 'Release' accordingly. For example:

```
    if (tasks.findByName("externalNativeBuildCustomDebug")) {
        copyCustomDebugAxslcToAssets.dependsOn externalNativeBuildCustomDebug
        compileCustomDebugJavaWithJavac.dependsOn copyCustomDebugAxslcToAssets
    }

    if (tasks.findByName("externalNativeBuildCustomRelease")) {
        copyCustomReleaseAxslcToAssets.dependsOn externalNativeBuildCustomRelease
        compileCustomReleaseJavaWithJavac.dependsOn copyCustomReleaseAxslcToAssets
    }

```
Does this make sense?

**Reply by rh101** [reply]
> Does this make sense?

Sure, assuming your variant names are "CustomDebug" and "CustomRelease"; is that correct?

Regardless though, at least now you have a solution to get it all working.

**Reply by murasu** [reply]
Absolutely! Thank you!

**Reply by halx99** [reply]
NEWS: the gradle script was improved: https://github.com/axmolengine/axmol/commit/dd7c9deaf9d0d837261ca6c43cba692474442be2
